### DISCUSSION TITLE
Switch between User presence (one for all guilds) and GuildMember (one per guild) presence modes

### DISCUSSION THREAD

**Discussion by kevinbioj** [discordjs/discord.js#5245] [General]
**Is your feature request related to a problem? Please describe.**
Right now, presences are sent for **each** GuildMember via `PRESENCE_UPDATE` and `GUILD_CREATE` events. The library handle them and cache them, very correct behaviour.
Unfortunately having one presence for each GuildMember makes the bot use too much memory, and if memory usage is a big requirement for you (example: having a low-sized RAM machine) it quickly become a serious problem.

Also, the `presence` getter of the `User` class take the first GuildMember presence associated to the member it finds by iterating over guilds ; since many developers use this getter they do not even get the current guild's presence which make the necessity of having GuildMember presences useless.

This could save RAM by having one presence per User and no longer one per GuildMember. I have also suggested an endpoint to fetch someone's presence on Discord API repo so we may just don't need to cache presences we don't need (and save **a lot** of RAM).

**Describe the ideal solution**
Have a `ClientOptions` property like `uniquePresences` which would be a boolean (default to `false` to respect Discord API scheme).
`true`: store all presences in `Client#presences` and leave `presences` of all `Guild` instances empty.
`false`: leave `Client#presences` empty (or not ?) and use `presences` of all `Guild` instances

**Describe alternatives you've considered**
If forking the repository so we can edit the code to make it store everything `Client#presences`, then yes. Otherwise, no.

**Additional context**
Before sweeping presences (~ 30 minutes of uptime w/ around 850 guilds and ~ 110k users):
```
- rss: 285.281MB
- heapTotal: 236.535MB
- heapUsed: 184.760MB
- external: 4.940MB
```

After:
```
- rss: 262.758MB
- heapTotal: 213.035MB
- heapUsed: 171.975MB
- external: 0.872MB
```

Longer you'll leave the bot online bigger the mem-delta will be.

**Comment by almostSouji** [comment]
I'm fully against this. While at the moment this seems to be sensible since the data should theoretically not vary based on guild this could change at any point if discord decides to do so, as per API specification presences are intrinsically bound to specific guilds.

**Reply by kevinbioj** [reply]
Since presences can now be completly ignored using intents, and being fetched on-demand... there is no reason to keep this discussion alive (as the main reason was memory consumption).
I'm not sure how to "close" a discussion on GitHub, if it ever can be closed.
